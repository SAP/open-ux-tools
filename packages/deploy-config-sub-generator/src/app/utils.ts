import type { DeployConfigOptions } from '../types';

/**
 * Parses the target from the CLI args or the options.
 *
 * @param args - cli args
 * @param opts - options
 * @returns - the target
 */
export function parseTarget(args: string | string[], opts: DeployConfigOptions): string | undefined {
    let result: string | undefined;
    if (typeof args === 'string') {
        result = args;
    } else if (Array.isArray(args)) {
        result = args?.[0];
    }
    if (!result) {
        result = opts.target;
    }
    return result;
}
