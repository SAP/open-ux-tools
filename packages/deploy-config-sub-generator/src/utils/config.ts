import { join } from 'path';
import { DeployProjectType } from '@sap-ux/abap-deploy-config-sub-generator';
import { UI5Config } from '@sap-ux/ui5-config';
import { FileName } from '@sap-ux/project-access';
import type { FioriToolsProxyConfigBackend } from '@sap-ux/ui5-config';
import type { Editor } from 'mem-fs-editor';
import type { DeployConfigOptions } from '../types';

/**
 * Retrieves backend configuration from either the base config (ui5.yaml) or from the options passed in.
 *
 * @param fs - file system editor
 * @param options - options passed in
 * @param launchStandaloneFromYui - flag to indicate if this generator is launched in YUI standalone
 * @param projectRoot - project root
 * @returns - backend configuration
 */
export async function getBackendConfig(
    fs: Editor,
    options: DeployConfigOptions,
    launchStandaloneFromYui: boolean,
    projectRoot: string
): Promise<{ backendConfig: FioriToolsProxyConfigBackend; isLibrary: boolean }> {
    let backendConfig: FioriToolsProxyConfigBackend;
    let isLibrary = false;
    // This is called when this generator is called as a subgenerator from
    // application generator or application modeler launcher (i.e. this.launchDeployConfigAsSubGenerator === true).
    if (launchStandaloneFromYui) {
        // Launched from app modeler where deploy config might already exist
        // need to retrieve backendConfig information.
        const ui5Config = await UI5Config.newInstance(fs.read(join(projectRoot, options.base ?? FileName.Ui5Yaml)));
        backendConfig = ui5Config.getBackendConfigsFromFioriToolsProxyMiddleware()[0];
        isLibrary = ui5Config.getType() === DeployProjectType.Library;
    } else {
        // Launched as subgenerator from app gen
        backendConfig = {
            destination: options.appGenDestination || options.connectedSystem?.destination?.Name,
            url: options.appGenServiceHost,
            client: options.appGenClient,
            scp: !!options.connectedSystem?.backendSystem?.serviceKeys || false
        } as FioriToolsProxyConfigBackend;
    }
    return { backendConfig, isLibrary };
}
