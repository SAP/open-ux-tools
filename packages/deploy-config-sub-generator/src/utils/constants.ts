// When deployment generator is bundled the namespacing is relative to the root generator
import { TargetName } from '@sap-ux/deploy-config-generator-shared';
import type { Target } from '../types';

export const generatorNamespace = (bundledRootGeneratorName: string, subGenName: string): string =>
    `${bundledRootGeneratorName}_${subGenName}`;

export const abapChoice: Target = { name: TargetName.ABAP, description: 'ABAP' };
export const cfChoice: Target = { name: TargetName.CF, description: 'Cloud Foundry' };

export const generatorTitle = 'Deployment Configuration Generator';
