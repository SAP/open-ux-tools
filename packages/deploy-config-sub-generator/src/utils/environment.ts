// Legacy package, dependent on external dependencies for async operations and no 'type: module' defined in package.json
import hasbin = require('hasbin');
import {
    API_BUSINESS_HUB_ENTERPRISE_PREFIX,
    ApiHubType,
    loadManifest,
    type ApiHubConfig
} from '@sap-ux/cf-deploy-config-sub-generator';
import {
    ERROR_TYPE,
    ErrorHandler,
    generateDestinationName,
    mtaExecutable
} from '@sap-ux/deploy-config-generator-shared';
import { existsSync } from 'fs';
import { join } from 'path';
import type { Editor } from 'mem-fs-editor';

/**
 * Check if the MTA is installed.
 *
 * @param choice - the choice (CF or ABAP)
 * @param projectPath - path to the project
 * @returns - true if the MTA is installed, otherwise an error message
 */
export function isMTAInstalled(choice: string, projectPath: string): boolean | string {
    if (
        (choice === 'cf' && !hasbin.sync(mtaExecutable)) ||
        (choice === 'abap' && !hasbin.sync(mtaExecutable) && existsSync(join(projectPath, 'mta.yaml')))
    ) {
        ErrorHandler.getErrorMsgFromType(ERROR_TYPE.NO_MTA_BIN);
        return ' ';
    }
    return true;
}

/**
 * Get the Api Hub Enterprise Key value from the node env if available.
 *
 * @returns The api hub enterprise config or undefined if the key is not found.
 */
export function getEnvApiHubConfig(): ApiHubConfig | undefined {
    const apiHubKey = process.env['API_HUB_API_KEY'];
    const apiHubType = process.env['API_HUB_TYPE'];

    // Legacy apps .env file will not define a type variable
    return apiHubKey
        ? {
              apiHubKey,
              apiHubType: apiHubType === ApiHubType.apiHubEnterprise ? ApiHubType.apiHubEnterprise : ApiHubType.apiHub
          }
        : undefined;
}

/**
 * Returns the destination name for API Hub Enterprise.
 *
 * @param memFs - reference to a mem-fs editor
 * @param opts -options representing the project app path and service path
 * @param opts.appPath - path to project
 * @param opts.servicePath - service path
 * @param apiHubConfig - API Hub Config
 * @returns - destination name
 */
export async function getApiHubOptions(
    memFs: Editor,
    { appPath, servicePath }: { appPath: string; servicePath: string | undefined },
    apiHubConfig?: ApiHubConfig
): Promise<{ destinationName: string | undefined; servicePath: string | undefined }> {
    let destinationName: string | undefined;
    if (apiHubConfig?.apiHubType === ApiHubType.apiHubEnterprise) {
        // appGenDestination may not have been passed in options e.g. launched from app modeler
        if (!servicePath) {
            // Load service path from manifest.json file
            const manifest = await loadManifest(memFs, appPath);
            servicePath = manifest?.['sap.app'].dataSources?.mainService?.uri;
        }
        destinationName = generateDestinationName(API_BUSINESS_HUB_ENTERPRISE_PREFIX, servicePath);
    }
    return { destinationName, servicePath };
}
