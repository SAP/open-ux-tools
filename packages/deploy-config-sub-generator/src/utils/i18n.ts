import i18next from 'i18next';
import translations from '../translations/deploy-config-sub-generator.i18n.json';
import { addi18nResourceBundle as addInquirerCommonTexts } from '@sap-ux/inquirer-common';
import type { i18n as i18nNext, TOptions } from 'i18next';

const deployConfigSubGen = 'deploy-config-sub-generator';
export const i18n: i18nNext = i18next.createInstance();

/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18n.init({
        lng: 'en',
        fallbackLng: 'en'
    });
    i18n.addResourceBundle('en', deployConfigSubGen, translations);

    addInquirerCommonTexts();
}

/**
 * Helper function facading the call to i18next. Unless a namespace option is provided the local namespace will be used.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    if (!options?.ns) {
        options = Object.assign(options ?? {}, { ns: deployConfigSubGen });
    }
    return i18n.t(key, options);
}

initI18n().catch(() => {
    // Needed for lint
});
