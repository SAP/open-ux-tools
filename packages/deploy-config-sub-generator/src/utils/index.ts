export * from './constants';
export * from './i18n';
export * from './environment';
