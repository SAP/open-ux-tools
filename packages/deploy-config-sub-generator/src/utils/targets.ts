import { DeployProjectType } from '@sap-ux/abap-deploy-config-sub-generator';
import { ApiHubType } from '@sap-ux/cf-deploy-config-sub-generator';
import { FileName } from '@sap-ux/project-access';
import { UI5Config } from '@sap-ux/ui5-config';
import { join } from 'path';
import { cfChoice, abapChoice } from './constants';
import type { Editor } from 'mem-fs-editor';
import type { Target } from '../types';
import type { ApiHubConfig } from '@sap-ux/cf-deploy-config-sub-generator';

/**
 * Generate a list of targets i.e. CF | ABAP and order based on the project type i.e. library, CF, abap or CAP.
 *
 * @param fs - reference to a mem-fs editor
 * @param projectPath - project path
 * @param isCap - is the target project a CAP project
 * @param hasMtaConfig - does the target project contain MTA Config
 * @param apiHubConfig - API Hub configuration
 * @param configFile - config file to read UI5 properties, default to ui5.yaml
 * @returns a list of Target options i.e. CF | ABAP
 */
export async function getSupportedTargets(
    fs: Editor,
    projectPath: string,
    isCap = false,
    hasMtaConfig = false,
    apiHubConfig?: ApiHubConfig,
    configFile = FileName.Ui5Yaml
): Promise<Target[]> {
    const isApiHubEnt = apiHubConfig?.apiHubType === ApiHubType.apiHubEnterprise;
    const isProjectExtension = fs.exists(join(projectPath, '.extconfig.json'));
    const ui5Config = await UI5Config.newInstance(fs.read(join(projectPath, configFile)));

    if (isApiHubEnt || isCap) {
        return [cfChoice];
    } else if (ui5Config.getType() === DeployProjectType.Library || isProjectExtension) {
        return [abapChoice]; // Extension projects, Library and systems using Reentrance tickets for auth
    } else {
        // If there's an mta.yaml in the hierarchy, it's probably a CF project
        // Offer that first and let the user decide
        return hasMtaConfig ? [cfChoice, abapChoice] : [abapChoice, cfChoice];
    }
}
