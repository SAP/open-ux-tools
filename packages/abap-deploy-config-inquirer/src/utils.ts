import { isAppStudio, listDestinations } from '@sap-ux/btp-utils';
import { getService } from '@sap-ux/store';
import { PackageInputChoices, TargetSystemType, TransportChoices } from './types';
import { getTransportConfigInstance } from './service-provider-utils';
import { t } from './i18n';
import LoggerHelper from './logger-helper';
import { listPackages } from './validator-utils';
import type {
    AbapDeployConfigAnswers,
    AbapDeployConfigAnswersInternal,
    BackendTarget,
    Credentials,
    InitTransportConfigResult,
    SystemConfig
} from './types';
import type { BackendSystem, BackendSystemKey } from '@sap-ux/store';
import type { Destinations, Destination } from '@sap-ux/btp-utils';
import { CREATE_TR_DURING_DEPLOY } from './constants';

let cachedDestinations: Destinations = {};
let cachedBackendSystems: BackendSystem[] = [];

/**
 * Retrieve the ABAP systems from the store or the destinations from BTP.
 *
 * @returns destinations or backend systems from the store
 */
export async function getAbapSystems(): Promise<{
    destinations: Destinations | undefined;
    backendSystems: BackendSystem[] | undefined;
}> {
    let destinations;
    let backendSystems;

    if (isAppStudio()) {
        destinations = await listDestinations({
            stripS4HCApiHosts: true
        });
        cachedDestinations = destinations;
    } else {
        const systemStore = await getService<BackendSystem, BackendSystemKey>({
            logger: LoggerHelper.logger,
            entityName: 'system'
        });
        backendSystems = await systemStore.getAll();
        cachedBackendSystems = backendSystems;
    }

    return { destinations, backendSystems };
}

/**
 * Retrieve a specific Destination, based on the destination name.
 *
 * @param destination - destination name
 * @returns destination if found
 */
export function findDestination(destination: string): Destination | undefined {
    return cachedDestinations?.[destination];
}

/**
 * Retrieve a specific BackendSystem, based on the URL.
 *
 * @param backendUrl - backend system URL
 * @returns backend system if found
 */
export function findBackendSystemByUrl(backendUrl: string): BackendSystem | undefined {
    return cachedBackendSystems?.find((backend: BackendSystem) => backend.url === backendUrl);
}

/**
 * Check if the current system is the same as the one in the answers.
 *
 * @param abapSystem - system configuration
 * @param url - url
 * @param client - client
 * @param destination - destination
 * @returns true if the system is the same
 */
export function isSameSystem(abapSystem?: SystemConfig, url?: string, client?: string, destination?: string): boolean {
    return Boolean(
        (abapSystem?.url &&
            abapSystem.url.trim()?.replace(/\/$/, '') === url?.trim()?.replace(/\/$/, '') &&
            abapSystem.client === client) ??
            (!!abapSystem?.destination && destination === abapSystem?.destination)
    );
}

/**
 * Get transport configuration from the backend.
 *
 * @param transportConfigParams - transport configuration parameters
 * @param transportConfigParams.backendTarget - backend target from prompt options
 * @param transportConfigParams.scp - scp
 * @param transportConfigParams.url - url
 * @param transportConfigParams.client - client
 * @param transportConfigParams.destination - destination
 * @param transportConfigParams.errorHandler - error handler
 * @param transportConfigParams.credentials - user credentials
 * @returns transport configuration
 */
export async function initTransportConfig({
    backendTarget,
    scp,
    url,
    client,
    destination,
    credentials,
    errorHandler
}: {
    backendTarget?: BackendTarget;
    scp?: boolean;
    url?: string;
    client?: string;
    destination?: string;
    credentials?: Credentials;
    errorHandler: (errorMessage: string) => void;
}): Promise<InitTransportConfigResult> {
    let result: InitTransportConfigResult = {};
    if (!url && !destination) {
        return result;
    }

    const systemConfig = {
        url,
        client,
        destination
    };

    try {
        result = await getTransportConfigInstance({
            backendTarget,
            scp,
            credentials,
            systemConfig
        });
    } catch (e) {
        result.error = e;
    }

    if (result.error) {
        errorHandler(result.error);
        LoggerHelper.logger.debug(
            t('errors.debugAbapTargetSystem', { method: 'initTransportConfig', error: result.error })
        );
    }

    return result;
}

/**
 * Querying package names that match the user input.
 *
 * @param input - user input
 * @param inputSystemConfig System configuration extracted from user answers for establishing backend connection
 * @param backendTarget - backend target from abap deploy config prompt options
 * @returns list of package names
 */
export async function queryPackages(
    input: string,
    inputSystemConfig: SystemConfig,
    backendTarget?: BackendTarget
): Promise<string[]> {
    const uppercaseInput = (input ?? '').toUpperCase();
    return listPackages(uppercaseInput, inputSystemConfig, backendTarget);
}

/**
 * Determines the package from the various package related prompts.
 *
 * @param previousAnswers - previous answers
 * @param statePackage - package from state
 * @returns package name
 */
export function getPackageAnswer(previousAnswers?: AbapDeployConfigAnswersInternal, statePackage?: string): string {
    // Older versions of YUI do not have a packageInputChoice question
    return statePackage || previousAnswers?.packageInputChoice === PackageInputChoices.ListExistingChoice
        ? previousAnswers?.packageAutocomplete ?? ''
        : previousAnswers?.packageManual ?? '';
}

/**
 * Determines the transport request from the various transport related prompts.
 *
 * @param promptAnswers - previous answers
 * @returns transport request
 */
export function getTransportAnswer(promptAnswers?: AbapDeployConfigAnswersInternal): string {
    return (
        promptAnswers?.transportManual ||
        promptAnswers?.transportFromList ||
        promptAnswers?.transportCreated ||
        (promptAnswers?.transportInputChoice === TransportChoices.CreateDuringDeployChoice
            ? CREATE_TR_DURING_DEPLOY
            : '')
    );
}

/**
 * Check if the transport matches placeholder used to create transport request number during actual deploy process.
 *
 * @param transport - existing transport
 * @returns true if transport setting is set to 'CreateDuringDeployChoice'.
 */
export function useCreateTrDuringDeploy(transport?: string): boolean {
    return transport === CREATE_TR_DURING_DEPLOY;
}

/**
 * Determines the url from the various sources.
 *
 * @param answers - internal abap deploy config answers
 * @param stateUrl - url from state
 * @returns url if found
 */
function getUrlAnswer(answers: AbapDeployConfigAnswersInternal, stateUrl?: string): string {
    let url = answers.url;
    if (stateUrl) {
        url = stateUrl;
    }
    return url;
}

/**
 * Convert internal answers to external answers to be used for writing deploy config.
 *
 * @param answers - internal abap deploy config answers, direct from prompting
 * @param state - partial internal abap deploy config answers derived from the state
 * @returns - external abap deploy config answers
 */
export function reconcileAnswers(
    answers: AbapDeployConfigAnswersInternal,
    state: Partial<AbapDeployConfigAnswersInternal>
): AbapDeployConfigAnswers {
    // Add dervied service answers to the answers object
    answers = Object.assign(answers, state);

    const reconciledAnswers: AbapDeployConfigAnswers = {
        url: getUrlAnswer(answers, state.url),
        package: getPackageAnswer(answers, state.package)
    };

    if (answers.destination) {
        reconciledAnswers.destination = answers.destination;
    }

    if (answers.targetSystem && answers.targetSystem !== TargetSystemType.Url) {
        reconciledAnswers.url = answers.targetSystem;
    }

    if (answers.client || state.client) {
        reconciledAnswers.client = answers.client || state.client;
    }

    if (answers.scp || state.scp) {
        reconciledAnswers.scp = true;
    }

    if (answers.ui5AbapRepo) {
        reconciledAnswers.ui5AbapRepo = answers.ui5AbapRepo;
    }

    if (answers.description) {
        reconciledAnswers.description = answers.description;
    }

    const transportAnswer = getTransportAnswer(answers);
    if (transportAnswer) {
        reconciledAnswers.transport = transportAnswer;
    }

    if (answers.index !== undefined) {
        reconciledAnswers.index = answers.index;
    }

    if (answers.overwrite !== undefined) {
        reconciledAnswers.overwrite = answers.overwrite;
    }

    return reconciledAnswers;
}
