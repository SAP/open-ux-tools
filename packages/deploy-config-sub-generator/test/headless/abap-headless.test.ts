import { join } from 'path';
import * as childProcess from 'child_process';
import { copy } from 'fs-extra';
import { readFile } from 'fs/promises';
import { rimraf } from 'rimraf';
import { runHeadlessGen } from './utils';
import { DeployTarget, type TelemetryHelper, hostEnvironment } from '@sap-ux/fiori-generator-shared';
import {
    backendSystemBtp,
    backendSystemOnPrem,
    INPUT_APP_DIR_ABAP,
    INPUT_BASE_APP,
    mockDestinations
} from './fixtures/constants';
import { getService } from '@sap-ux/store';
import { isAppStudio, listDestinations } from '@sap-ux/btp-utils';

jest.mock('@sap-ux/fiori-generator-shared', () => {
    return {
        ...(jest.requireActual('@sap-ux/fiori-generator-shared') as {}),
        sendTelemetry: jest.fn(),
        TelemetryHelper: {
            initTelemetrySettings: jest.fn(),
            createTelemetryData: jest.fn()
        } as TelemetryHelper,
        getHostEnvironment: () => {
            return hostEnvironment.cli;
        }
    };
});

jest.mock('@sap-ux/store', () => {
    return {
        ...(jest.requireActual('@sap-ux/store') as {}),
        getService: jest.fn()
    };
});

jest.mock('@sap-ux/btp-utils', () => {
    return {
        ...(jest.requireActual('@sap-ux/btp-utils') as {}),
        isAppStudio: jest.fn(),
        listDestinations: jest.fn()
    };
});

export const ORIGINAL_CWD: string = process.cwd(); // Generators change the cwd, this breaks sonar report so we restore later
export const OUTPUT_DIR = join(__dirname, '../test-output/abap');

jest.mock('child_process');
let spawnMock: jest.SpyInstance;

const getServiceMock = getService as jest.Mock;
const isAppStudioMock = isAppStudio as jest.Mock;
const listDestinationsMock = listDestinations as jest.Mock;

describe('Test ABAP headless generator', () => {
    beforeAll(async () => {
        rimraf.sync(OUTPUT_DIR);
        await copy(INPUT_APP_DIR_ABAP, OUTPUT_DIR);
        // This is a hack to ensure it only returns CLI in all situations
        process.stdin.isTTY = true;
    });

    beforeEach(() => {
        jest.resetAllMocks();
        spawnMock = jest.spyOn(childProcess, 'spawnSync').mockImplementation(() => ({ status: 0 } as any));
    });
    afterEach(() => {
        process.chdir(ORIGINAL_CWD);
    });

    afterAll(() => {
        // Remove the test folder if the folder is empty (i.e. no failed tests)
        try {
            // if (readdirSync(OUTPUT_DIR).length === 0) {
            //     console.log('Removing test output folder');
            //     rimraf.sync(OUTPUT_DIR);
            // }
        } catch {
            // do nothing
        }
    });

    it('should generate ABAP deploy config for on-prem system (VSCode)', async () => {
        isAppStudioMock.mockReturnValue(false);
        getServiceMock.mockResolvedValue({
            getAll: jest.fn().mockResolvedValue([backendSystemOnPrem])
        });
        const headlessConfig = 'app1';
        const testAppName = 'app-on-prem';

        await copy(join(OUTPUT_DIR, INPUT_BASE_APP), join(OUTPUT_DIR, testAppName));
        await runHeadlessGen(headlessConfig, DeployTarget.ABAP, join(OUTPUT_DIR, testAppName));

        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/package.json`, 'utf-8')).toMatchSnapshot();
        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/ui5-deploy.yaml`, 'utf-8')).toMatchSnapshot();
    });

    it('should generate ABAP deploy config for btp system (VSCode)', async () => {
        isAppStudioMock.mockReturnValue(false);
        getServiceMock.mockResolvedValue({
            getAll: jest.fn().mockResolvedValue([backendSystemBtp])
        });
        const headlessConfig = 'app1';
        const testAppName = 'app1-btp';

        await copy(join(OUTPUT_DIR, INPUT_BASE_APP), join(OUTPUT_DIR, testAppName));
        await runHeadlessGen(headlessConfig, DeployTarget.ABAP, join(OUTPUT_DIR, testAppName));

        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/package.json`, 'utf-8')).toMatchSnapshot();
        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/ui5-deploy.yaml`, 'utf-8')).toMatchSnapshot();
    });

    it('should generate ABAP deploy config for an on-prem destination (BAS)', async () => {
        isAppStudioMock.mockReturnValue(true);
        listDestinationsMock.mockResolvedValue(mockDestinations);
        const headlessConfig = 'app2-bas';
        const testAppName = 'app2-bas-on-prem';

        await copy(join(OUTPUT_DIR, INPUT_BASE_APP), join(OUTPUT_DIR, testAppName));
        await runHeadlessGen(headlessConfig, DeployTarget.ABAP, join(OUTPUT_DIR, testAppName));

        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/package.json`, 'utf-8')).toMatchSnapshot();
        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/ui5-deploy.yaml`, 'utf-8')).toMatchSnapshot();
    });

    it('should generate ABAP deploy config for an s4hc destination (BAS)', async () => {
        isAppStudioMock.mockReturnValue(true);
        listDestinationsMock.mockResolvedValue(mockDestinations);
        const headlessConfig = 'app3-bas';
        const testAppName = 'app3-bas-s4hc';

        await copy(join(OUTPUT_DIR, INPUT_BASE_APP), join(OUTPUT_DIR, testAppName));
        await runHeadlessGen(headlessConfig, DeployTarget.ABAP, join(OUTPUT_DIR, testAppName));

        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/package.json`, 'utf-8')).toMatchSnapshot();
        expect(await readFile(`${OUTPUT_DIR}/${testAppName}/ui5-deploy.yaml`, 'utf-8')).toMatchSnapshot();
    });
});
