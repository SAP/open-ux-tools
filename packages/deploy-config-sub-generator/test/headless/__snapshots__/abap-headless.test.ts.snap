// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test ABAP headless generator should generate ABAP deploy config for an on-prem destination (BAS) 1`] = `
"{
  "name": "baseapp",
  "version": "0.0.1",
  "description": "An SAP Fiori application.",
  "keywords": [
    "ui5",
    "openui5",
    "sapui5"
  ],
  "main": "webapp/index.html",
  "dependencies": {},
  "devDependencies": {
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "rimraf": "^5.0.5"
  },
  "scripts": {
    "start": "fiori run --open \\"test/flp.html#app-preview\\"",
    "start-local": "fiori run --config ./ui5-local.yaml --open \\"test/flp.html#app-preview\\"",
    "build": "ui5 build --config=ui5.yaml --clean-dest --dest dist",
    "deploy": "npm run build && fiori deploy --config ui5-deploy.yaml",
    "deploy-config": "fiori add deploy-config",
    "start-noflp": "fiori run --open \\"index.html?sap-ui-xx-viewCache=false\\"",
    "int-test": "fiori run --open 'test/integration/opaTests.qunit.html'",
    "start-variants-management": "fiori run --open \\"preview.html#app-preview\\"",
    "unit-test": "fiori run --open 'test/unit/unitTests.qunit.html'",
    "undeploy": "npm run build && fiori undeploy --config ui5-deploy.yaml",
    "deploy-test": "npm run build && fiori deploy --config ui5-deploy.yaml --testMode true"
  },
  "sapuxLayer": "CUSTOMER_BASE"
}
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for an on-prem destination (BAS) 2`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: "3.1"
metadata:
  name: baseapp
type: application
builder:
  resources:
    excludes:
      - /test/**
      - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          destination: Dest1
          url: https://mock.url.dest1.com
        app:
          name: MY_UI5_ABAP_REPO
          description: Testing headless ABAP deployment on BAS
          package: ZPKG123
          transport: $TMP
        exclude:
          - /test/
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for an s4hc destination (BAS) 1`] = `
"{
  "name": "baseapp",
  "version": "0.0.1",
  "description": "An SAP Fiori application.",
  "keywords": [
    "ui5",
    "openui5",
    "sapui5"
  ],
  "main": "webapp/index.html",
  "dependencies": {},
  "devDependencies": {
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "rimraf": "^5.0.5"
  },
  "scripts": {
    "start": "fiori run --open \\"test/flp.html#app-preview\\"",
    "start-local": "fiori run --config ./ui5-local.yaml --open \\"test/flp.html#app-preview\\"",
    "build": "ui5 build --config=ui5.yaml --clean-dest --dest dist",
    "deploy": "npm run build && fiori deploy --config ui5-deploy.yaml",
    "deploy-config": "fiori add deploy-config",
    "start-noflp": "fiori run --open \\"index.html?sap-ui-xx-viewCache=false\\"",
    "int-test": "fiori run --open 'test/integration/opaTests.qunit.html'",
    "start-variants-management": "fiori run --open \\"preview.html#app-preview\\"",
    "unit-test": "fiori run --open 'test/unit/unitTests.qunit.html'",
    "undeploy": "npm run build && fiori undeploy --config ui5-deploy.yaml",
    "deploy-test": "npm run build && fiori deploy --config ui5-deploy.yaml --testMode true"
  },
  "sapuxLayer": "CUSTOMER_BASE"
}
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for an s4hc destination (BAS) 2`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: "3.1"
metadata:
  name: baseapp
type: application
builder:
  resources:
    excludes:
      - /test/**
      - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          destination: Dest2
          url: https://mock.s4hana.url.dest2.com
          authenticationType: reentranceTicket # SAML support for vscode
        app:
          name: MY_UI5_ABAP_REPO
          description: Headless ABAP deployment for S4HC
          package: ZPKG123
          transport: TEST_TR
        exclude:
          - /test/
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for btp system (VSCode) 1`] = `
"{
  "name": "baseapp",
  "version": "0.0.1",
  "description": "An SAP Fiori application.",
  "keywords": [
    "ui5",
    "openui5",
    "sapui5"
  ],
  "main": "webapp/index.html",
  "dependencies": {},
  "devDependencies": {
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "rimraf": "^5.0.5"
  },
  "scripts": {
    "start": "fiori run --open \\"test/flp.html#app-preview\\"",
    "start-local": "fiori run --config ./ui5-local.yaml --open \\"test/flp.html#app-preview\\"",
    "build": "ui5 build --config=ui5.yaml --clean-dest --dest dist",
    "deploy": "npm run build && fiori deploy --config ui5-deploy.yaml",
    "deploy-config": "fiori add deploy-config",
    "start-noflp": "fiori run --open \\"index.html?sap-ui-xx-viewCache=false\\"",
    "int-test": "fiori run --open 'test/integration/opaTests.qunit.html'",
    "start-variants-management": "fiori run --open \\"preview.html#app-preview\\"",
    "unit-test": "fiori run --open 'test/unit/unitTests.qunit.html'",
    "undeploy": "npm run build && fiori undeploy --config ui5-deploy.yaml",
    "deploy-test": "npm run build && fiori deploy --config ui5-deploy.yaml --testMode true"
  },
  "sapuxLayer": "CUSTOMER_BASE"
}
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for btp system (VSCode) 2`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: "3.1"
metadata:
  name: baseapp
type: application
builder:
  resources:
    excludes:
      - /test/**
      - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          url: https://example.abap.backend:44300
          client: '100'
        app:
          name: ZTESTREPO
          description: Testing headless ABAP deployment
          package: ZPKG12
          transport: TR123
        exclude:
          - /test/
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for on-prem system (VSCode) 1`] = `
"{
  "name": "baseapp",
  "version": "0.0.1",
  "description": "An SAP Fiori application.",
  "keywords": [
    "ui5",
    "openui5",
    "sapui5"
  ],
  "main": "webapp/index.html",
  "dependencies": {},
  "devDependencies": {
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "rimraf": "^5.0.5"
  },
  "scripts": {
    "start": "fiori run --open \\"test/flp.html#app-preview\\"",
    "start-local": "fiori run --config ./ui5-local.yaml --open \\"test/flp.html#app-preview\\"",
    "build": "ui5 build --config=ui5.yaml --clean-dest --dest dist",
    "deploy": "npm run build && fiori deploy --config ui5-deploy.yaml",
    "deploy-config": "fiori add deploy-config",
    "start-noflp": "fiori run --open \\"index.html?sap-ui-xx-viewCache=false\\"",
    "int-test": "fiori run --open 'test/integration/opaTests.qunit.html'",
    "start-variants-management": "fiori run --open \\"preview.html#app-preview\\"",
    "unit-test": "fiori run --open 'test/unit/unitTests.qunit.html'",
    "undeploy": "npm run build && fiori undeploy --config ui5-deploy.yaml",
    "deploy-test": "npm run build && fiori deploy --config ui5-deploy.yaml --testMode true"
  },
  "sapuxLayer": "CUSTOMER_BASE"
}
"
`;

exports[`Test ABAP headless generator should generate ABAP deploy config for on-prem system (VSCode) 2`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: "3.1"
metadata:
  name: baseapp
type: application
builder:
  resources:
    excludes:
      - /test/**
      - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          url: https://example.abap.backend:44300
          client: '100'
        app:
          name: ZTESTREPO
          description: Testing headless ABAP deployment
          package: ZPKG12
          transport: TR123
        exclude:
          - /test/
"
`;
